drivers/usb/udc/udc_mcux.c

```diff
@@ -0,0 +1,903 @@
+static const usb_device_controller_interface_struct_t mcux_ip3511_usb_iface = {
+	USB_DeviceLpc3511IpInit, USB_DeviceLpc3511IpDeinit, USB_DeviceLpc3511IpSend,
+	USB_DeviceLpc3511IpRecv, USB_DeviceLpc3511IpCancel, USB_DeviceLpc3511IpControl
+};
+#endif
+
+struct udc_mcux_config {
+	const usb_device_controller_interface_struct_t *mcux_if;
+	void (*irq_enable_func)(const struct device *dev);
+	void (*irq_disable_func)(const struct device *dev);
+	size_t num_of_eps;
+	struct udc_ep_config *ep_cfg_in;
+	struct udc_ep_config *ep_cfg_out;
+	uint32_t base;
+	const struct pinctrl_dev_config *pincfg;
+#ifdef CONFIG_USB_UDC_NXP_PHY
+	uint8_t need_to_init_phy;
+	uint8_t D_CAL;
+	uint8_t TXCAL45DP;
+	uint8_t TXCAL45DM;
+#endif
+};
+
+struct udc_mcux_data {
+	const struct device *dev;
+	usb_device_struct_t device_instance;
+	uint8_t controller_id;
+};
+
+/* TODO: implement the cache maintenance
+ * solution1: Use the non-cached buf to do memcpy before/after giving buffer to usb controller.
+ * solution2: Use cache API to flush/invalid cache. but it needs the given buffer is
+ * cache line size aligned and the buffer range cover multiple of cache line size block.
+ * Need to change the usb stack to implement it, will try to implement it later.
+ */
+#if defined(CONFIG_NOCACHE_MEMORY)
+K_HEAP_DEFINE_NOCACHE(mcux_packet_alloc_pool, USB_DEVICE_CONFIG_ENDPOINTS * 2u * 1024u);
+
+/* allocate non-cached buffer for usb */
+static void *udc_mcux_nocache_alloc(uint32_t size)
+{
+	void *p = (void *)k_heap_alloc(&mcux_packet_alloc_pool, size, K_NO_WAIT);
+
+	if (p != NULL) {
+		(void)memset(p, 0, size);
+	}
+
+	return p;
+}
+
+/* free the allocated non-cached buffer */
+static void udc_mcux_nocache_free(void *p)
+{
+	if (p == NULL) {
+		return;
+	}
+	k_heap_free(&mcux_packet_alloc_pool, p);
+}
+#endif
+
+static int usb_mcux_control(const struct device *dev, usb_device_control_type_t command,
+				void *param)
+{
+	const struct udc_mcux_config *config = dev->config;
+	struct udc_mcux_data *priv = udc_get_private(dev);
+	usb_status_t status;
+
+	status = config->mcux_if->deviceControl(priv->device_instance.controllerHandle,
+			command, param);
+
+	if (status != kStatus_USB_Success) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+/* If ep is busy, return busy. Otherwise feed the buf to controller */
+static int usb_mcux_ep_feed(const struct device *dev,
+			struct udc_ep_config *const cfg,
+			struct net_buf *const buf)
+{
+	const struct udc_mcux_config *config = dev->config;
+	struct udc_mcux_data *priv = udc_get_private(dev);
+	usb_status_t status = kStatus_USB_Success;
+	uint8_t *data;
+	uint32_t len;
+	int current_sr;
+	usb_device_endpoint_status_struct_t ep_status;
+
+	ep_status.endpointAddress = cfg->addr;
+	usb_mcux_control(dev, kUSB_DeviceControlGetEndpointStatus, &ep_status);
+	if (ep_status.endpointStatus == kUSB_DeviceEndpointStateStalled) {
+		return -EACCES; /* stalled */
+	}
+
+	current_sr = irq_lock();
+	if (!udc_ep_is_busy(dev, cfg->addr)) {
```

Comment:
I am using the `busy `state to determine whether to prime next transfer to controller. I think there are two contexts that will prime transfers to controller. (1) `USB_DeviceNotificationTrigger` is called in isr when one transfer is done, then it will check queued transfer and prime it to controller; (2) application or other threads can queue transfer and prime it to controller too. So I need the `irq_lock` because I don't use the `work` as udc_kinetis.c did.